cpp/src/arrow/compute/exec/asof_join_node.cc

```diff
@@ -140,6 +172,17 @@ struct MemoStore {
   // Stores last known values for all the keys
 
   struct Entry {
+    Entry() = default;
+
+    Entry(OnType time, std::shared_ptr<arrow::RecordBatch> batch, row_index_t row)
+        : time(time), batch(batch), row(row) {}
+
+    void swap(Entry& other) {
```

Comment:
We generally discourage mutable references but given this is supposed to be an analogue of `std::swap` it seems ok.